### ISSUE TITLE
_handle_config() throws error on certain data values - 1.13.9

### ISSUE THREAD

**Issue by ajschmidt8** [nutechsoftware/alarmdecoder#47]
I discovered this bug in the process of attempting to upgrade Home Assistant to use AlarmDecoder `1.13.9` (we're currently stuck [on 1.13.2](https://github.com/home-assistant/core/blob/dev/homeassistant/components/alarmdecoder/manifest.json#L5) with the pinned `pyserial` and `futures` dependencies).

It looks like there are some invalid values being passed to the `_handle_config()` function that causes AlarmDecoder to crash with the following error:
```
ValueError: too many values to unpack (expected 2)
```

I captured a screenshot from the Python debugger below

![ad_stacktrace](https://user-images.githubusercontent.com/7400326/79700287-a1223180-8262-11ea-955e-a2e871515bde.png)

The problematic `data` value is:

```
!CONFIG>MODE=A&CONFIGBITS=ff05&ADDRESS=19&LRR=N&COM=N&EXP=NNNNN&REL=NNNN&MASK=ffffffff&DEDUPLICATE=N\r\n!Reading configuration.\r\n!UART init.\r\n!VER:ffffffff,V2.2a.8.8,TX;RX;SM;VZ;RF;ZX;RE;AU;3X;CG;DD;MF;L2;KE;M2;CB;DS;ER;CR\r\n!>\r\n!CONFIG>MODE=A&CONFIGBITS=ff05&ADDRESS=19&LRR=N&COM=N&EXP=NNNNN&REL=NNNN&MASK=ffffffff&DEDUPLICATE=N\r\n!Reading configuration.\r\n!UART init.\r\n!VER:ffffffff,V2.2a.8.8,TX;RX;SM;VZ;RF;ZX;RE;AU;3X;CG;DD;MF;L2;KE;M2;CB;DS;ER;CR\r\n!AUI:1606000000006363024543f531fb456cf5ec01010201
```

When you run [`data.split('>')`](https://github.com/nutechsoftware/alarmdecoder/blob/master/alarmdecoder/decoder.py#L572) on the above value it returns 4 values instead of 2, thus causing the error:

![image](https://user-images.githubusercontent.com/7400326/79700383-332a3a00-8263-11ea-9804-2850035572d3.png)

**Comment by ajschmidt8** [comment]
upon inspection, it looks like this could be related to https://github.com/nutechsoftware/alarmdecoder/issues/40.

**Comment by f34rdotcom** [comment]
I need more people testing this patch but ya that seems like it may be the issue. Only thing I noted is it looks like it is not parsing out the \r\n and splitting the message by lines so when it does the split it gets very confused.

**Comment by f34rdotcom** [comment]
Merged #41 to dev branch for testing.

**Comment by ajschmidt8** [comment]
awesome. i'll try to do some testing later this week. appreciate it @f34rdotcom.

**Comment by ajschmidt8** [comment]
@f34rdotcom, I can confirm that the fix from https://github.com/nutechsoftware/alarmdecoder/pull/41 did not resolve the problem I faced. I'm still seeing the error message below.

![image](https://user-images.githubusercontent.com/7400326/80316324-34131c80-87cb-11ea-809b-228df924fe13.png)

**Comment by ajschmidt8** [comment]
@f34rdotcom, any updates on this? is there anything I can do to help expedite a fix here?

**Comment by ajschmidt8** [comment]
@f34rdotcom, checking in again. Do you have any updates on this issue? It's currently a blocker for us that's preventing us from updating our AlarmDecoder dependency in Home Assistant.

**Comment by f34rdotcom** [comment]
Lost track of this thread. I will do my best to stay on this but its clearly not a simple fix. 

I have read this thread again and it looks like a parsing problem where lines are not split on '\r\n' in some instances and IMHO the serial reading routine needs help. It should be digesting a single character at a time. This uses a fraction more CPU but also avoids complex algorithms to to deal with splitting lines.

I am going to roll back this last patch on 'dev' and do some work in dev to rewrite the serial reading routine over the next few days while I wait for SmartThings dev team to fix a few issues holding back my other work on the embedded AD2 version I am currently working on.

**Comment by f34rdotcom** [comment]
@ajschmidt8 and others
  Please test the latest commit(d809cb3) in the dev branch. I am testing under my docker version at home and so far no issues.

**Comment by ajschmidt8** [comment]
> @ajschmidt8 and others
> Please test the latest commit([d809cb3](https://github.com/nutechsoftware/alarmdecoder/commit/d809cb34cd6123088aa29e94f7b40897f7e7b40a)) in the dev branch. I am testing under my docker version at home and so far no issues.
> 
> Note:
> I am using a Rock64 with docker so I am likely not under powered and not a good test candidate for this exact issue but any general parsing issues I should have seen by now.

thanks! I will test these changes this weekend. appreciate the quick turnaround. I'll let you know how it goes.

**Comment by ajschmidt8** [comment]
@f34rdotcom, I tested these changes with Home Assistant. I seem to be getting the error message shown in the screenshot below, `TypeError: 'int' object is not subscriptable`. There's a longer stack trace here, but it looks like it mostly just repeats what's in the screenshot: https://pastebin.com/hQ6hipTw

Hopefully this helps. Let me know if there's any additional information I can provide to help troubleshoot this. Appreciate your work on this.


![image](https://user-images.githubusercontent.com/7400326/91673776-b148f800-eb03-11ea-97d8-584d5a28e697.png)

**Comment by ajschmidt8** [comment]
@f34rdotcom, I posted some changes in #56 that seem to have fixed the issue described in my screenshot above.

**Comment by kerryb48** [comment]
@f34rdotcom  FYI - I've been running the dev branch with @ajschmidt8's fixes with my homeassistant instance for several days and it seems to be stable and has resolved my issue

**Comment by ajschmidt8** [comment]
resolved in #56. closing
